Version 0.3.1 (#22)
LLNL/elastic-stacker: docs/usage.md

# Elastic Stacker

## User manual

Elastic Stacker is a tool for importing and exporting configuration data from
Elasticsearch and Kibana instances (collectively an "*Elastic Stack*").

It can process:

- Indices
- Index templates
- Component templates
- Ingest Pipelines
- Enrich policies
- Transforms
- Watches
- Roles
- Role mappings
- Saved Objects, including but not limited to:
  - Dashboards
  - Visualizations
  - Saved queries and searches
  - Lenses
  - Index patterns

It can dump (but not load):

- Agent policies
- Package policies

These can be added to the list once the Elastic Agent APIs come out of their
alpha state and stabilize enough to work with.

It cannot currently process:

- the actual contents of indices
- ILM policies

(There is no technical reason Stacker can't handle these; it just hasn't been

> done yet...if you want to contribute these features see contributing.md to get

Reviewer: ```suggestion
done yet...if you want to contribute these features see [the contributing docs](docs/contributing.md) to get
```

started.)

This document will walk you through how to use Stacker to back up and restore
configuration data, or to move it between multiple stacks.

## Configuration

Stacker adjusts its behavior based on a configuration file, called

> `stacker.yaml` (the `a` in yaml is optional). It looks for this file in your

Reviewer: I think I've actually been standardizing on `.yml` but it's fine.

current working directory, in your `$XDG_CONFIG_HOME` (this is `~/.config` on
Linux), and finally in `/etc`.

### Defaults

The top-level key in the config file is `default` -- the options in the
`default` section will be used unless overriden somewhere else. If you only need
to configure Stacker one way, it's perfectly acceptable to put all your
configuration in this section. Here's a list of configuration keys that can go
in the defaults section:

#### `client`

The client section configures the base HTTP client used for communication
with Elasticsearch and Kibana. You can set options like:

- `client.base_url`: The URL for the Elasticsearch or Kibana APIs. This should
  usually not be set in the `client` section, but instead overridden in the

>   `elasticsearch` and `kibana` sections that apply to just one or the other.

Reviewer: Not sure if I understand this wording?

Author: Fixed in 8dfbe0ac65d. Arguably the `client` section shouldn't even accept a `base_url` parameter since there's no way both Elasticsearch and Kibana could be listening on the same host and port, so you're right that it shouldn't be documented in the `client` section.

- `client.timeout`: The timeout for requests, in seconds
- `client.verify`: the path to a file containing CA certificates which the
  client will trust.
- `client.auth`: a username and password to use for HTTP Basic authentication --
  set `auth.username` and `auth.password` to authenticate this way.
- `client.tls`: authentication using mutual TLS. Set `tls.cert` and `tls.key` as
  paths to the certificate and key files to use for authentication.
- `client.headers`: A map of HTTP headers to set on the request and the values
  they should take. If you want to use API keys instead of a username and
  password to interact with your Elastic Stack, you should set it as a HTTP

>   header like so:
>   ```yaml

Reviewer: ```suggestion
  header like so:
  
  ```yaml
```

Author: Fixed in ccb66987477

>   header like so:
>   ```yaml
>   client:
>     headers:
>       Authorization: Apikey <key>
>   ```

Reviewer: ```suggestion
  header like so:

  ```yaml
  client:
    headers:
      Authorization: Apikey <key>
  ```
```

   `search` string in single quotes.

1. Substitutions are applied to each file in lexicographic order based on the
   name. This means that as the profile is merged with the defaults, the
   defaults may be applied before or after the profile. For example:

   ```yaml
   default:
     substitutions:
       01_one:
       03_three:
   profiles:
     staging:
       substitutions:
         00_zero:
         02_two:
   ```

   In this situation, the substitutions would be applied to your dump files in
   order, 00-03.

#### `options`

This section holds miscellaneous global options for the Stacker application.
Currently supported keys are:

- `data_directory`: Where to write out the files dumped from Elasticsearch.
- `watcher_users`: A mapping of `username: password` pairs for users in the
  Elastic [Watcher](https://www.elastic.co/guide/en/kibana/current/watcher-ui.html).
  The Elasticsearch API censors passwords from Watcher users by default, so if
  you want to load dumped watchers into another cluster you need to provide the
  password for the Watcher user separately.

#### `log`:

- `log.level`: sets the global log level for the application; accepts values

>   like "INFO" or "warning"

Reviewer: Does case matter here?

Author: UPPER- or lowercase are both acceptable (but mixed-case is not.)
Clarified in f768785a.

> - `log.ecs`: formats the log messages in [Elastic Common

Reviewer: I assume this takes true / false values?

Author: Correct; clarified in 2b477c6f3

```yaml
default:
  client:
    timeout: 10

profiles:
  prod:
    client:
      timeout: 30
```

In this situation, Stacker will set the client timeout to 10 seconds, unless it
is run with the `--profile prod` flag in which case it will set the timeout to
30 seconds.

It is perfectly acceptable to set options in a profile which you haven't set in
the defaults; the options will be merged together when Stacker is run.

> :information: Options given at the command line, like `--data-directory`
> override both the defaults and the profile.

## Command Line Usage

This section describes how to use Stacker's command-line flags effectively.

### Subcommands

The Stacker command requires a subcommand. You can see a list of all of them by
running `stacker` with no arguments. Most of these subcommands work the same
way: they have a `load` and a `dump` subcommand which exports or imports that
type of data into the configured data directory. For example:

```bash
stacker pipelines dump
stacker saved-objects load
```

These subcommands accept a number of flags which will be discussed more in the
next section.

There are a few other subcommands which behave differently:

> - `system-dump`and `system-load`: iterate over all types of resources Stacker knows

Reviewer: ```suggestion
- `system-dump` and `system-load`: iterate over all types of resources Stacker knows
```

  about, and dump or load all of them.
- `profile`: show the current configuration profile, after all the various
  levels of configuration have been merged together.
- `version`: Show the installed version of the `elastic-stacker` package.

### Flags and Parameters

In general, the individual resource subcommands (like `transforms` or
`component-templates`) accept most of the same parameters. If you want to see

> what flags a particular subcommand takes, you can run `stacker --help <subcommand>` to see what flags it takes. (One exception to this is the `load`

Reviewer: This is helpful, I've found it hard to intuit the help structures of the various commands the way my brain expects them. E.g. my brain wants to do `stacker <subcommand> --help` but that leads to a stack trace IIRC.

I assume this is just the way that `fire` works for handling these?

Author: This is partially an issue with `fire` and partially a flawed design decision on my part -- all the commands that lead to a stack trace are ones that accept `**kwargs`, and Fire interprets *anything* after the subcommand as a keyword argument, even `--help` as a keyword argument to pass into the function.

Those functions need `**kwargs` right now because not all of them accept the same flags, and the `system_dump` and `system_load` methods currently call all the individual load and dump methods with the same arguments -- this is an area of future work.

> and `dump` subcommands -- these do not respect the help flag. This is a known

Reviewer: Is there a link available to an issue for this "known" bug?

Author: https://github.com/LLNL/elastic-stacker/issues/23

- `--include-experimental`: Dump resources which cannot yet be loaded back in
  (agent policies and package policies.)

## Tips and Tricks

This section describes some recommended uses for Stacker, and how to go about
them effectively.

### Using Stacker to back up your Elastic configuration

Stacker dumps all the data it can handle as JSON files in a defined directory
structure. The easiest way to store data like this is in a Git repository.

Create a new repo, and in the root of the repo place a `stacker.yaml` file
which sets `defaults.options.data_directory: "."`. Now, while you're in this
directory, Stacker will use the configuration file in that directory first, and
dumped files will be put directly in the repo. With this setup in place, you can
back up your data by running `stacker system-dump`, then review the changes and
commit them.

### Using Stacker to move configuration between Elastic Stacks

Stacker can be used to keep configuration in sync between several different
Elastic Stacks (for example, a staging and a production environment.)

To configure Stacker for this use case, you may want to start with

> [`stacker.example.yaml`](../stacker.example.yaml) which shows how this can work.

Reviewer: ```suggestion
[`stacker.example.yaml`](/stacker.example.yaml) which shows how this can work.
```

I think you can do this to use absolute path from the base of the repo.

> In that configuration, there's one profile for `staging` (preproduction) and one

Reviewer: ```suggestion
In that configuration, there's one profile for `staging` (pre-production) and one
```
